### ISSUE TITLE
[P2] 犬・猫画像分類の学習〜推論デモを実装する

### ISSUE THREAD

**Issue by tomiya7688** [tomiya7688/comfyUI_support_tools#96]
## 目的
数学的シミュレーション／AI学習デモの一部として、Kaggle等でよく見られる **犬 vs 猫の画像分類デモ**を実装する。

単に分類結果を出すだけではなく、

- 画像を読み込む
- 前処理する
- モデルへ入力する
- 学習する
- loss / accuracy が変化する
- 学習済みモデルで犬/猫を推論する

という一連の流れを視覚的に理解できることを目的とする。

## 実装言語に関する制約
**アプリ本体の実装言語・既存アーキテクチャは変更しない。**

Pythonは以下の補助用途では利用してよい。
- モデル学習
- 推論処理
- 学習済みモデルの生成
- 前処理スクリプト
- 評価スクリプト

ただし、UIやアプリ全体をPythonへ移行しない。

既存アプリからPython処理を呼ぶ場合は、明確なbridge / process boundary / API boundaryを設ける。

---

## 画像・データセットのライセンス方針
**デモで使用・同梱・再配布する犬・猫画像は、商用/非商用を問わず再利用・再配布が明確に許可された完全フリー素材のみを使用する。**

要件:
- [ ] 画像ごとに出典とライセンスを確認する
- [ ] 再配布不可・用途制限付き・出典不明の画像は同梱しない
- [ ] Kaggle由来画像は、個々の画像ライセンスが明確に確認できない限り同梱しない
- [ ] 可能なら Public Domain / CC0 を最優先する
- [ ] CC BY等を利用する場合は、必要な帰属表示をリポジトリ内に保持する
- [ ] `assets/licenses/` 等に出典・作者・ライセンス・元URL等を記録する
- [ ] 学習用データセットを自動取得する場合も、取得先の利用条件を確認する
- [ ] テストやREADMEのスクリーンショットに表示する画像も同じ基準を満たす

「ネット上で無料閲覧できる画像」と「自由に再配布できる画像」を混同しない。

---

### 1. データセット
犬・猫の2クラス画像を使用する。

方針:
- Public Domain / CC0等の明確に再配布可能な画像を優先
- ライセンス確認済みのフリー素材だけで最小データセットを構成
- 開発用・デモ用に同梱する画像も同じ基準を適用

本IssueではKaggle自体への依存は必須にしない。むしろライセンスが曖昧なKaggle由来画像はデモ同梱用途では避ける。

### 2. 学習
理解しやすさを優先し、初期実装は小規模CNNまたは軽量なtransfer learningでよい。

例:
```text
image
  ↓
resize / normalize
  ↓
CNN / feature extractor
  ↓
classifier
  ↓
cat probability / dog probability
```

### 3. 推論
任意画像を入力し、以下を表示する。
- Dog probability
- Cat probability
- 最終判定
- confidence

---

## 実装内容
- [ ] 犬・猫分類デモ用ページを追加
- [ ] 画像を1枚選択して推論できる
- [ ] 画像プレビューを表示
- [ ] 推論結果を `Dog / Cat` で表示
- [ ] クラス確率をbar等で表示
- [ ] 学習済みモデルを読み込んで推論できる
- [ ] 学習デモモードを用意
- [ ] epochごとのlossを表示
- [ ] epochごとのaccuracyを表示
- [ ] validation loss / accuracyも可能なら表示
- [ ] 学習中の進捗を表示
- [ ] 学習後に同じサンプルを再推論し、結果変化を確認できる
- [ ] モデル未学習 / 学習済みの比較表示を可能にする

---

## 視覚デモ
ページ内で最低限以下の流れを見せる。

```text
Image
  ↓
Resize / Normalize
  ↓
Feature extraction
  ↓
Classifier
  ↓
Dog / Cat probability
```

学習時:
```text
Prediction
  ↓
Compare with label
  ↓
Loss
  ↓
Update weights
  ↓
Next epoch
```

#92 の機械学習学習ループデモと対応づける。

---

## UIイメージ
```text
[ Input Image ]
   cat.jpg

[ Preprocess ]
  resize: 128x128
  normalize: on

[ Model ]
  simple CNN

[ Prediction ]
  Cat: 0.91
  Dog: 0.09

Result: Cat
```

学習画面:
```text
Epoch 1/10
loss: 0.68
accuracy: 0.61

Epoch 5/10
loss: 0.39
accuracy: 0.84

Epoch 10/10
loss: 0.21
accuracy: 0.92
```

loss / accuracyはグラフでも表示する。

---

## Python利用方針
Pythonは別プロセスまたは明確な補助moduleとして扱う。

候補構成:
```text
Main Application
  ↓
ML demo adapter
  ↓
Python process / CLI / local service
  ↓
train.py / infer.py
```

または、モデルをONNX等へexportできる場合は、

```text
Python training
  ↓
model export
  ↓
Main application inference runtime
```

としてもよい。

重要:
- アプリ本体の言語は変更しない
- Python依存がない環境でも他の数学デモが壊れない
- MLデモ機能だけoptional dependencyとして扱える構造を優先する

---

## モデル候補
MVPでは精度より理解しやすさと軽量性を優先する。

候補:
- 小規模CNN
- MobileNet系transfer learning
- EfficientNetの小型variant

実装者が最も依存を増やさず扱いやすいものを選択してよい。

---

## 特徴量の視覚化
#91 の「特徴量抽出デモ」とつながるよう、可能であれば中間層の出力を簡易表示する。

- [ ] early convolution feature mapを数枚表示
- [ ] edge / textureのような反応があることを説明
- [ ] 最終的な高次元特徴の意味を断定しない

説明例:
- 「初期層では輪郭や模様のような局所パターンへの反応が見えます」
- 「後段では複数の特徴を組み合わせて分類します」

この表示は教育用デモであり、feature attributionの厳密な説明を目的としない。

---

## データ分割
最低限:
- train
- validation

を分ける。

可能ならtestも分離する。

UI上では簡単に以下を説明する。
- train: 学習に使う
- validation: 学習途中の確認に使う
- test: 最終評価に使う

---

## 評価表示
- [ ] accuracy
- [ ] loss
- [ ] confusion matrix（可能なら）
- [ ] 犬を猫と誤判定した例
- [ ] 猫を犬と誤判定した例

「正解率100%を目標にする」デモではなく、誤分類例も見せて機械学習が完全ではないことを確認できるようにする。

---

## 再現性
- [ ] random seedを固定できる
- [ ] 使用model / epoch / learning rate等を記録
- [ ] 同じ設定で概ね再現可能にする

---

## テスト
- [ ] 学習済みモデルをロードできる
- [ ] 犬画像1枚から2クラス確率を返せる
- [ ] 猫画像1枚から2クラス確率を返せる
- [ ] 確率の合計が概ね1になる
- [ ] 学習実行でloss historyが生成される
- [ ] accuracy historyが生成される
- [ ] UIがPython未導入時にアプリ全体をクラッシュさせない
- [ ] MLデモ以外の既存機能はPython依存なしで利用できる
- [ ] 同梱する全画像について出典・ライセンス情報が記録されている
- [ ] 再配布条件を満たさない画像がリポジトリへ含まれていない

---

## 完了条件
- 犬/猫画像を入力して分類結果と確率を表示できる
- 学習から推論までの流れを視覚的に追える
- loss / accuracyの変化を確認できる
- 誤分類例を確認できる
- Pythonを利用してもアプリ本体の実装言語を変更しない
- Python側処理との境界が明確になっている
- #92 の学習ループで説明した概念が実画像分類でも確認できる
- デモに同梱・表示する画像がすべて再配布可能なフリー素材である
- 画像ライセンス情報をリポジトリ内で追跡できる

## 関連
- #91 セル・オートマトンの特徴量抽出デモ
- #92 機械学習の学習ループを視覚的に説明するデモページ
- #94 微分・積分・偏微分・連鎖律を視覚化する数学デモページ

## Out of Scope
- 大規模モデル学習
- Kaggle leaderboard参加
- ライセンス不明なKaggle画像の同梱
- 犬種 / 猫種の多クラス分類
- 物体検出 / segmentation
- GPU必須化
- アプリ本体のPython化
- クラウド学習基盤
- 本番用途の画像認識サービス
